pallene/types.lua

```diff
@@ -66,27 +83,43 @@ function types.equals(t1, t2)
         end
 
         for i = 1, #t1.ret_types do
-            if not types.equals(t1.ret_types[i], t2.ret_types[i]) then
+            if not equivalent(t1.ret_types[i], t2.ret_types[i], is_gradual) then
                 return false
             end
         end
 
         return true
-    elseif tag1 == tag2 then
-        return true
+
+    elseif tag1 == types.T.Record then
+        return t1.type_decl == t2.type_decl
+
+    elseif tag1 == types.T.Builtin then
+        return t1.builtin_decl == t2.builtin_decl
+
     else
-        return false
+        return error("impossible")
     end
 end
 
+function types.equals(t1, t2)
+    return equivalent(t1, t2, false)
+end
+
+-- The type-consistency relation from gradual typing.
+-- It is equality relaxed with (T ~ value) and (value ~ T)
+function types.consistent(t1, t2)
+    return equivalent(t1, t2, true)
+end
+
 function types.tostring(t)
     local tag = t._tag
-    if     tag == types.T.Integer     then return "integer"
-    elseif tag == types.T.Boolean     then return "boolean"
-    elseif tag == types.T.String      then return "string"
+    if     tag == types.T.Value       then return "value"
```

Comment:
We should start using alphabetic ordering for this kind of elseif chains.

Reply:
I'd rather make a big global change like this in a separate PR. Should we create a new issue for it? (I personally don't mind it that much. The types enumeration is loosely grouped based on primitive vs non-primitive)

Reply:
This is minor, nevermind.